File: 03-java-socket/src/main/java/demo/ProtelisConfigSpec.java
package demo;

import com.uchuhimo.konf.ConfigSpec;
import com.uchuhimo.konf.RequiredItem;
import demo.data.ProtelisNode;

import java.util.List;

/**
 * Konf specification for the TOML configuration file.
 */
public final class ProtelisConfigSpec {

    /**
     * Create a new config specification named protelis.
     */
    public static final ConfigSpec SPEC = new ConfigSpec("protelis");
    /**
     * Required field iterations. It represents the number of cycles the simulation will perform.
     */
    @SuppressWarnings("PMD.FieldNamingConventions")
    // CHECKSTYLE: ConstantName OFF
    public static final RequiredItem<Integer> iterations =
            new RequiredItem<>(SPEC, "iterations") { };
    /**
     * Required field protelisModuleName. It is the name of the .pt source file.
     */
    @SuppressWarnings("PMD.FieldNamingConventions")
    // CHECKSTYLE: ConstantName OFF
    public static final RequiredItem<String> protelisModuleName =
            new RequiredItem<String>(SPEC, "protelisModuleName") { };
    /**
     * Required field nodes. It contains a list of protelis nodes.
     */
    @SuppressWarnings("PMD.FieldNamingConventions")
    // CHECKSTYLE: ConstantName OFF
    public static final RequiredItem<List<ProtelisNode>> nodes =
            new RequiredItem<List<ProtelisNode>>(SPEC, "nodes") { };

    private ProtelisConfigSpec() { }
}

File: 04-kotlin-socket/src/main/kotlin/ProtelisConfigSpec.kt
package demo

import com.uchuhimo.konf.ConfigSpec

data class IPv4Host(val host: String, val port: Int)
data class ProtelisNode(val hostandport: IPv4Host, val id: Int, val neighbors: Set<IPv4Host>, val leader: Boolean = false)

object ProtelisConfigSpec : ConfigSpec("protelis") {
    val iterations by required<Int>(description = "Number of iterations the nodes have to run.")
    val protelisModuleName by required<String>()
    val nodes by required<List<ProtelisNode>>(description = "List of nodes")
}
